maven-core/src/main/java/org/apache/maven/ReactorReader.java

```diff
@@ -263,15 +271,28 @@ private boolean isPackagedArtifactUpToDate( MavenProject project, File packagedA
             while ( iterator.hasNext() )
             {
                 Path outputFile = iterator.next();
+
+                if ( Files.isDirectory(  outputFile ) )
+                {
+                    continue;
+                }
+
                 long outputFileLastModified = Files.getLastModifiedTime( outputFile ).toMillis();
                 if ( outputFileLastModified > artifactLastModified )
                 {
-                    LOGGER.warn(
-                            "Packaged artifact for {} is not up-to-date compared to the build output directory; "
-                          + "file {} is more recent than {}.",
-                            project.getArtifactId(),
-                            relativizeOutputFile( outputFile ), relativizeOutputFile( packagedArtifactFile.toPath() )
-                    );
+                    File alternative = determineLooseDirectoryForArtifact( project, artifact );
+                    if ( alternative != null )
+                    {
+                        LOGGER.warn( "File {} is more recent than the packaged artifact for {}; using {} instead",
+                                relativizeOutputFile( outputFile ), project.getArtifactId(),
+                                relativizeOutputFile( alternative.toPath() ) );
+                    }
+                    else
+                    {
+                        LOGGER.warn( "File {} is more recent than the packaged artifact for {}; "
+                                + "cannot use a loose directory for this type of artifact",
```

Comment:
I have a bit of a problem with this: where is the defition of a "loose directory"?

Reply:
There's no formal definition of a "loose directory" - the term is taken from [this comment](https://github.com/apache/maven/blob/8eda091d4e2231eee3c33b4e9d417d62971d0762/maven-core/src/main/java/org/apache/maven/ReactorReader.java#L187).

I think it is important to let the user know that the requested artifact could not be resolved from the workspace. How about this?

> ... cannot use the build output directory for this type of artifact

Reply:
This sounds better, but introduces another term 'workspace' which is in Maven Core, but not visible to users too.

Reply:
Updated, let me know what you think.

Reply:
WFM